gen/internals.py

```diff
@@ -184,28 +212,131 @@ def add_entry(self, entry, replace_existing):
         self.add_conditional_scope(entry, [])
 
 
-# NOTE: This exception should never escape the DFSArgumentCalculator
+# NOTE: This exception should never escape the Resolver
 class CalculatorError(Exception):
-    def __init__(self, message, chain=[]):
-        assert isinstance(message, str)
+
+    def __init__(self, message: str, chain=None):
+        if chain is None:
+            chain = list()
         assert isinstance(chain, list)
         self.message = message
         self.chain = chain
         super().__init__(message)
 
 
+class SkipError(Exception):
+    """Raised when this key should silently produce no errors."""
+    pass
+
+
+class Resolvable:
+    """ Keeps track of the state of the resolution of a variable with a given name
+
+    - Manages the states (unresolved vs. error / success)
+    - Keeps track of what ended up setting the value
+    - Keeps track of what caused the resolvable to enter existence (it's cause)
+    """
+
+    UNRESOLVED = 'unresolved'
+    ERROR = 'error'
+    RESOLVED = 'resolved'
+
+    def __init__(self, name, cause):
+        self._state = Resolvable.UNRESOLVED
+        self.name = name
+        self.cause = cause
+        self.error = None
+        self.setter = None
+        self._value = None
+
+    @property
+    def is_finalized(self):
+        return self._state != Resolvable.UNRESOLVED
+
+    @property
+    def is_error(self):
+        return self._state == Resolvable.ERROR
+
+    @property
+    def is_resolved(self):
+        return self._state == Resolvable.RESOLVED
+
+    def finalize_error(self, error):
+        assert self._state == Resolvable.UNRESOLVED
+        self._state = 'error'
+        self.error = error
+
+    def finalize_value(self, value: str, setter: Setter):
+        assert self._state == 'unresolved'
+        self._state = 'resolved'
+        self._value = value
+        self.setter = setter
+
+    @property
+    def value(self):
+        assert self._state == 'resolved', "is_resolved() must return true for this function to " \
+            "be called. State: {}, is_resolved: {}".format(self._state, self.is_resolved)
+
+        return self._value
+
+    def __str__(self):
+        return "<Resolvable name: {}, state: {}>".format(self.name, self._state)
+
+
+def yield_implicit_validates(target):
+    # Recursively walk the target / sub scope tree and yield a
+    # validate function for each and every switch.
+    for name, scope in target.sub_scopes.items():
+        # Note: It's important scope.cases.keys() is evaluated now, not later
+        # (is evaluated before the validate is called)
+        yield name, partial(validate_one_of, valid_values=scope.cases.keys())
+        for sub_target in scope.cases.values():
+            yield from yield_implicit_validates(sub_target)
+
+
+class ArgumentSet(dict):
```

Comment:
Practically this is trying _really_ hard to act more like a set than a dictionary (The key and the value have the same identity always, the key's identity is derived from the value being held). If I could I'd use a custom key function and wrap a set, but unfortunately python doesn't allow / have a way to do that.